### ISSUE TITLE
DelDOT TSDM - Structure/Conveyance Work Order Update

### ISSUE THREAD

**Issue by KCI-Ablowers** [leeensminger/DelDOT-NPDES-Viewer#338]
DelDOT has requested that the structure and conveyance work orders be updated to address multiple problem codes -

- Update Structure and Conveyance Tables or View with the new `PARENT` field.
- Create a process to generate the "parent" record with the "Multiple" problem code in structures and conveyance work order table or view.
- Update the nightly TDSM process were necessary.

The general idea for how this should be set up is displayed in the attached excel file that shows the current, manual workflow for BMP Inspections (upcoming issue/task). A general description of the "count" logic is below -

If count `KCIWONO` = 1 then no "parent" record is needed and `PARENT` is null/blank. If count 'KCIWONO` >1 then the number of records sent should be equal to the number of problem codes associated with the work order plus 1 (for the parent record). 

[1st_Quarter_BMP_WO_2019_04_02_19.xlsx](https://github.com/leeensminger/DelDOT-NPDES-Field-Tool/files/3740316/1st_Quarter_BMP_WO_2019_04_02_19.xlsx)

**Comment by gerrykelly** [comment]
reminders to self:
- replace base tables with evw views

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
Just adding the status comment that prototype for the TSDM product has been developed as views in the dev GDB and as pending review by "Newark".  Correspondence on this development was better handled by emails up to this point, but let's now bring it back into the GH issue.

**Comment by KCI-Ablowers** [comment]
@gerrykelly I spoke with @cstevenson1 today about pushing this to DelDOT. In her email on 11/5, she said - 

> Thanks Gerry. We are now seeing the 60 records with “Multiple” as Andrew referenced in his original email. The process appears to be working as described. Matt and I agree that it is good to be passed along to DelDOT for approval. If the plan is to pass the DEV table to Supreet, I would make sure he is aware of the data inaccuracies that should not be an issue in PROD (DEV has those extra work order records and the supervisors are outdated).

For us to pass this to DelDot, we thought of two options:

1. Export the DEV views to excel and state that some of the data will be missing or inaccurate because it is the DEV version.
2. Push the PROD data to DEV, then pass the view onto DelDOT so the changes we made in PROD are reflecting in the DEV database.

Which of these two do you think is the most efficient?

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
If the goal here is simply to give DelDOT something to review for structure and content -- with the dev difference from prod as a know issue, the export to Excel seems like the better short run option

**Comment by KCI-Ablowers** [comment]
@gerrykelly that works for me. I exported both the WORK_ORDERS_MS4_TSDM_temp and the WORK_ORDERS_MS4_withParentWO to Excel - 
[WORK_ORDERS_MS4_TSDM_temp.xlsx](https://github.com/leeensminger/DelDOT-NPDES-Viewer/files/3839118/WORK_ORDERS_MS4_TSDM_temp.xlsx)
[WORK_ORDERS_MS4_withParentWO.xlsx](https://github.com/leeensminger/DelDOT-NPDES-Viewer/files/3839119/WORK_ORDERS_MS4_withParentWO.xlsx)

I think we should send both along and see if they are interested in the feature_type or feature_num files from the "...withParentWO" table in the final "...MS4_TSDM" output. 

Attaching the files here so I can grab them again to email to the team.

**Comment by KCI-Ablowers** [comment]
I sent the information off to Supreet at DelDOT this afternoon for review.

**Comment by cstevenson1** [comment]
@KCI-Ablowers per Matt's conversation with Supreet:

1. "Drainage" should be populated for FAIL_CLASS on ALL records. This is currently working for child records, but needs to be implemented for parent records (where PROBLEM_CO = "Multiple")
2. PRIORITY should be populated for ALL records. This is currently working fine for child records but logic needs to be implement for parents records (where PROBLEM_CO = "Multiple"). If possible, Matt and I suggest looking at the associated child records and populating PRIORITY on the parent record based on the highest value of the associated child records. For example: If the parent has 2 child records and one has a PRIORITY = 3 and the other has a PRIORITY = 5, we would like the parent record to have PRIORITY = 5. 
3. FUNC_CO should be populated for all records. This is currently working fine for child records but logic needs to be implement for parents records (where PROBLEM_CO = "Multiple"). If possible, Matt and I suggest looking at the associated child records and populating FUNC_CO on the parent record based on the problem codes of the child records. If PROBLEM_CO = “OPFLOOD” OR “OPSTWATE” then populate 2199.0. For all others, populate 2299.0 in FUNC_CO on the parent record. 
4. As anticipated, Supreet would like to receive the photos associated with the work order. The defect photos appear to be located in the WORK_ORDER_PHOTOS table and Supreet would like them to be presented the same way as the current TSDM delivery that is already set up for structure and conveyance photos (with a link?)
5. Due to problem codes updates covered in number 6, "problem" will need to be populated. KCI Newark will handle this in the PROD database (in the work order comment field) prior to the initial submission and then going forward, it will be populated by field staff in the field tool. 
6. Currently, in D_ProblemCode, the description is the same as the code. It should not be this way. The values need to be updated according to the attached excel. Existing work order records will need to be updated to reflect these changes (KCI Newark can help with this part if needed). Not sure how this impacts the field tool, but two check boxes will need to be removed from the work order UI and three display values need to be modified.

[Maximo Problem Codes for Drainage_11_21_19.xlsx](https://github.com/leeensminger/DelDOT-NPDES-Viewer/files/3888151/Maximo.Problem.Codes.for.Drainage_11_21_19.xlsx)

![image](https://user-images.githubusercontent.com/16919898/69568780-4540ce80-0f8a-11ea-8d37-2ceef4dc9a6d.png)

**Comment by KCI-Ablowers** [comment]
@kanglu666 For #6 above https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-558294516 please review the excel and let me know if you have any questions. The updates to the work order UI include removing the two checkboxes highlighted in yellow and modifying the three values circled in red.

**Comment by KCI-Ablowers** [comment]
@cstevenson1 Per @kanglu666 the DB was designed with identical codes and descriptions from the beginning. From my understanding of item 6 above the description can still display but the problem code needs to write to the work order, is that correct?

**Comment by cstevenson1** [comment]
@KCI-Ablowers The identical codes and descriptions was not a design decision made by the Newark group but it has been that way for a while. This has been pushed to the back burner until now because work orders were not being submitted to DelDOT. It however was not like that in our previous database. But yes you are correct; the description can still display but the code needs to be written to the work order record.

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
From the 11/25/2019 comments, these issues are addressed as follows (in dev and prod):
- the view -- WORK_ORDERS_MS4_withParentWO will be the product for structure/conveyance WOs going forward (we can name it something better)
- For the FAIL_CLASS the assumption is that **all** WORK_ORDER_PROBLEM_CODES records related to a WO will have the same value for FAILURE_CLASS (currently "drainage" is the only available value).  If , for WORK_ORDER_PROBLEM_CODES records related to a singe WO, it is possible that there could be a mix of values or some with a value and some null, I would need logic to avoid creating quasi-duplicate records in the view.
- For FUNC_CO, the view checks for PROBLEM_CODE = “OPFLOOD” OR “OPSTWATE” and for the current values that match those ('Open Drainage Standing Water','Open Drainage Flooding') and sets the value to 2199.0 if either of those  PROBLEM_CO values are found in any of the child records .  I have the "For all others, populate 2299.0 in FUNC_CO" on the parent record doing so even if all child record PROBLEM_CODEs are NULL ... checking if that is correct, or if the parent record  FUNC_CO should instead be null in that case.

- The view now has a photos_link field that works the same as the existing field in the tables of the TSDM delivery to build a URL to get the BASE64 photos, like:

FC=WORK_ORDERS&GID=%7B25C237A7-61DF-40FC-9E5F-040D98A7A1EC%7D&f=pjson
....
https://newarkgis.kci.com/arcgis/rest/services/DelDOT_TSDM/FeatureDetailTool/GPServer/FeatureDetailTool/execute?**FC=WORK_ORDERS&GID=%7B25C237A7-61DF-40FC-9E5F-040D98A7A1EC%7D&f=pjson**&token=AZMG_y7ej0NoUoeRaQhhsgdsG5Vgsz-JZsJx-k_EQ9s5e5kzjG3CvblvSrFdDf4W

then from the response value, like:
"{'LATEST PHOTOS': {u'Defect_2': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B68CE4FE3-7457-4DCE-B14D-7DB276E0B13F%7D&f=pjson', u'Defect_1': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B95C383D4-5870-4A78-A02C-EFC87FAECAB3%7D&f=pjson', u'Defect': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B5A5EE96D-9C1E-448D-81F8-224920985D7E%7D&f=pjson'}}"

They would retrieve an HTML doc with the photo exactly as with the current TSDM scheme, like:
https://newarkgis.kci.com/arcgis/rest/services/DelDOT_TSDM/returnBASE64imageString/GPServer/returnBASE64imageString/execute?**attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B95C383D4-5870-4A78-A02C-EFC87FAECAB3%7D&f=pjson**&token=AZMG_y7ej0NoUoeRaQhhsgdsG5Vgsz-JZsJx-k_EQ9s5e5kzjG3CvblvSrFdDf4W

I have verified that the value then returned can be pasted into an HTML, like https://htmlcodeeditor.com/ , and the doc/photo is rendered as expected.

**Comment by KCI-Ablowers** [comment]
@gerrykelly 

## From Christina's email on 1/17/2019 - Struct/Convey and BMP WO Requests 

> b.	Likewise the .0 is dropped off at the end of some function codes.

It looks like this could be due to different data types between your view and the DB table in production. `FUNC_CO` in your view is (nvarchar(58), not null) and `FUNCTION_CODE` in `dbo.LU_WORK_ORDER_CODES` is (numeric(38,8),null)

> 2.	As discussed the other day, only imd_action = No work orders should be included in the Structure/Conveyance work order submission. 

I did a cursory check and it appears that KCIWONOs with the "multiple" problem code are either Yes or No, not a mix for `imd_action`. 

> 3.	Will lat and long be calculated before each deliverable? The values are currently 10000.

It looks like this is being set explicitly in your view `10000.00000001 AS lon, 10000.00000001 AS lat`. Do you know if these values were calculated previously? or do they calculate at some point during the push to DelDOT?

> 4.	It appears the same photo link is used for all records of a single work order (see KCIWONO CA532BF1-D918-468D-86A1-04573F43CC7F as an example). Does this mean the same photo would be included multiples times in a given submission?
     a.	How would the process handle work orders that have multiple defect photos? 

Not sure what the LOE is to string in multiple photos to the photo link or make sure photos don't duplicate. Do you think this will be a major effort?

**Comment by KCI-Ablowers** [comment]
@gerrykelly An update to the query - 

Only records with a `STATUS` = "Open" should be included/submitted. This field is only updated once DelDOT notifies Newark that the WO is closed.

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
So far as "b. Likewise the .0 is dropped off at the end of some function codes." ...
- There are no function code values in LU_WORK_ORDER_CODES that end in ".0", nor can there be, since it is type=double, meaning a .0 is not preserved
- The code actually come from the WORK_ORDER_PROBLEM_CODES table where function_code field is a text field that uses the D_FunctionCode domain.  While that domain has several entries with ".0" 
- However the WORK_ORDER_PROBLEM_CODES table (1188 records) has only one code ending in a ".0" (whether you search in ArcMap or SSMS)
- And that does show up in the view:
![image](https://user-images.githubusercontent.com/8259136/74976576-fb377f80-53f6-11ea-9a74-4c4af43e80a8.png)
- but Excel brought in that field as a number rather than text, so no ".0"
![image](https://user-images.githubusercontent.com/8259136/74976816-7a2cb800-53f7-11ea-93a8-e5d16014b9ff.png)

So, if it is thought there should be more  ".0" records that is a data issue that could be looked into, but is not a result of my SQL

**Comment by KCI-Ablowers** [comment]
@cstevenson1 Regarding the lat/long from the comment above https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-575685582 - 

Is this a requirement from DelDOT or a "nice to have"? I spoke with @gerrykelly and up until now the submit process has been explicitly SQL based. Calculating the X/Y will involve developing a geoprocess that will need to be added in, tested, etc. Please let us know your and Matt's thoughts.

**Comment by gerrykelly** [comment]
@KCI-Ablowers @cstevenson1 

So far as "same photo link is used for all records of a single work order ", keep in mind that link is not to a single photo.  It is used to build the URL that returns the JSON with date for **all**  records in WORK_ORDER_PHOTOS related to the WORK_ORDER -- in a second HTTP request 

It so happens that the WO with globalID CA532BF1-D918-468D-86A1-04573F43CC7F has only one photo.  But, using WO with globalID 7F1C061D-2DEE-4123-A1A7-0CD6A477985C, you would have a link value: FC=WORK_ORDERS&GID=%7B7F1C061D-2DEE-4123-A1A7-0CD6A477985C%7D&f=pjson
Which would be put into the URL: https://newarkgis.kci.com/arcgis/rest/services/DelDOT_TSDM/FeatureDetailTool/GPServer/FeatureDetailTool/execute?FC=WORK_ORDERS&GID=%7B7F1C061D-2DEE-4123-A1A7-0CD6A477985C%7D&f=pjson&token=AZMG_y7ej0NoUoeRaQhhsgdsG5Vgsz-JZsJx-k_EQ9s5e5kzjG3CvblvSrFdDf4W

.... and the relevant data returned from that URL is
   "value": "{'LATEST PHOTOS': {u'Defect_2': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B69325AAC-C7F3-44D1-A643-07734331A5D9%7D&f=pjson', u'Defect_3': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B9767DA65-1720-4713-BE89-7355F0215A5E%7D&f=pjson', u'Defect_1': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7BD3B86F44-AF0C-4281-8BBC-2E211790FDBE%7D&f=pjson', u'Defect_4': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7BA07F0F8F-2649-408C-9F9E-013A7EFCD4C1%7D&f=pjson', u'Defect_5': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B7FEFC030-30E3-432C-BD68-9E9B59D2FEB9%7D&f=pjson', u'Defect': 'attachmentTable=WORK_ORDER_PHOTOS__ATTACH&RELGLOBALID=%7B27067192-6AC1-4205-A3B3-C8AFE8B6DAF3%7D&f=pjson'}}"


So, as seen, that gives the info needed to get any of 6 photos related to that WO.

Yes, the same photo link is used for all records of a single work order.  But that link is for all related photos.  Since WO photos relate directly to the WO, and not the problem code and do not have any further differentiation, at this point I can't see that there is another option.

**Comment by gerrykelly** [comment]
@KCI-Ablowers @cstevenson1 
For recording whether workorders have been submitted, the most straightforward way would be to add a field to the workorder table.

But that would not work unless for WOs that are output to multiple records for the DelDOT detail, like below, are always all "submitted" or not.  Meaning that if it will be the case that for KCIWONO CA532BF1-D918-468D-86A1-04573F43CC7F, the record with func_co 2240.1 was submitted, but the record for func_co 2250 was not I would need to do something else 

![image](https://user-images.githubusercontent.com/8259136/75052440-9daa3e00-549d-11ea-83c8-f046faee3b90.png)

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
I added "imd_action = 'No'" to the [WORK_ORDERS_MS4_withParentWO] view (working in DEV only for now)
- since imd_action is directly from WORK_ORDERS.IMMEDIATE_ACTION, there could not be differing values for the same-[KCIWONO] records
- Unless otherwise noted, [WORK_ORDERS_MS4_withParentWO] will be the only view being maintained for this work from here on
-That means that so far the only change since the last submittal is that the overall query for the view now has a where clause of STATUS='OPEN' and imd_action = 'No'
- Adding "SUBMITTED" to that criteria is pending clarifaction
- If I have overlooked any other filtering criteria, let me know

**Comment by cstevenson1** [comment]
Regarding the dropping the 0 comment (https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-589298977)

We looked into Gerry's comment and observed the same thing. It sounds like it is a string in his environment, which is good. The work order function code field utilizes D_FunctionCode, which has the .0 at the end as it should. However, there is only 1 record in the database with the .0 at the end, which is wrong since all the function codes in D_FunctionCode have a decimal. Our staff does not have the ability to edit this in the field tool so we are not sure how this happened. We are also unsure what the LU_WORK_ORDERS_CODES table is being used for. I'm guessing this is not an issue Gerry will be addressing but we will need to:
1. Make sure everything is working properly with D_FunctionCode and the field tool/database
2. Update existing records to reflect the codes listed in D_FunctionCode

D_FunctionCode
![image](https://user-images.githubusercontent.com/16919898/75176827-e3b40b80-5702-11ea-9ac9-2cf997b5a471.png)

LU_WORK_ORDERS_CODES
![image](https://user-images.githubusercontent.com/16919898/75176847-f29abe00-5702-11ea-9b0b-1de01829bc69.png)

**Comment by cstevenson1** [comment]
Regarding the lat/long comment https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-589310450

Matt spoke with DelDOT and stated "I checked with Supreet yesterday afternoon and confirmed with him that latitude and longitude values DO NOT need to submitted with both the Structure/Conveyance work orders as well as the BMP work orders. 

Supreet stated that these values are already carried on the assets themselves (Structures, Conveyances, and BMPs features) in Maximo and that this is sufficient."

Since we are not populating them, should the fields be removed so it is not misleading?

**Comment by cstevenson1** [comment]
Regarding the photos comment https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-589725578

Matt spoke with Supreet and Supreet is good with the way asset photos are submitted so we foresee no issues with using the same methodology for work order photos as Gerry proposed.

**Comment by cstevenson1** [comment]
@KCI-Ablowers regarding my comment (https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-590463774)

Should we (Newark) update the FUNCTION_CODE field? We are still unsure why they don't match the coded value domain.

Some records appear to be displaying the code and others the description for the problem code.
![image](https://user-images.githubusercontent.com/16919898/76422043-f7a27300-637a-11ea-8c4c-cf0099e9fca1.png)

**Comment by KCI-Ablowers** [comment]
@cstevenson1 did you mean the D_ProblemCode domain? They appear to be updated based on conversations with Lu and looking at the domain in Prod - 

![image](https://user-images.githubusercontent.com/35460336/76426369-59fe7200-6381-11ea-844b-57dee93d226c.png)

I believe we decided earlier on in this request that you guys would be responsible for updating the problem codes to match the new domain values.

**Comment by cstevenson1** [comment]
@KCI-Ablowers No I mean function code.

https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-589298977

https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-590463774

If there are no concerns with mass updating the records in DEV for testing, I can field calculate the problem codes and function codes. I know the problem codes were updated but were the function codes? The values are correct in the domain but incorrect in the database for records from this year so the past couple of months.

![image](https://user-images.githubusercontent.com/16919898/76428040-86b38900-6383-11ea-94be-b1cdb36fa847.png)

**Comment by KCI-Ablowers** [comment]
@cstevenson1 I don't see any harm in doing the field calculate in DEV. @gerrykelly Do you see any issues arising with that?

**Comment by gerrykelly** [comment]
@KCI-Ablowers @cstevenson1 
I don't see it as a risk ... and if it turns out otherwise, that's what dev is for.

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
Outstanding issues are addressed and updated in dev (both vm-deldotsql\deldot and deldot-sql2016) and prod
    The view WORK_ORDERS_MS4_withParentWO no longer exists and has been replaced by [WORK_ORDERS_MS4_Complete]

I'll assign this back to you for now,

**Comment by cstevenson1** [comment]
@gerrykelly 
Lat/Long should be removed
![image](https://user-images.githubusercontent.com/16919898/79880362-e73fd800-83bd-11ea-9a14-e73f2b50db52.png)

@KCI-Ablowers 
Any luck on figuring out why the function codes do not match up? I field calculated in Dev for testing purposes and will need to do the same thing in PROD. However, it still appears that all of the coded values for that domain have a value after the decimal but the ones ending in .0 are dropping the ".0" when they are used in the field tool and subsequently added in the database. This would require me to manually correct every work order problem code record that our staff creates that should end in .0

![image](https://user-images.githubusercontent.com/16919898/79883287-b95c9280-83c1-11ea-8ee3-10f2dd577d0b.png)

**Comment by KCI-Ablowers** [comment]
Contacted @kanglu666 this morning to pick back up on the "function code" issues noted above.

**Comment by KCI-Ablowers** [comment]
### Update on Fuction Code Issue

@kanglu666 was able to diagnose the problem and will continue working on updating the field tool, debugging, checking the DB, and releasing a new dev field tool version for testing.

**Comment by cstevenson1** [comment]
@KCI-Ablowers duplicate value present in field tool

![image](https://user-images.githubusercontent.com/16919898/84696098-f3599900-af19-11ea-8945-bdbfdfb6894a.png)

**Comment by cstevenson1** [comment]
decimal came through properly in the sync from Dev 9.3

![image](https://user-images.githubusercontent.com/16919898/85284991-188c6100-b45e-11ea-9d74-89890d70de06.png)

**Comment by KCI-Ablowers** [comment]
@kanglu666 - Can you remove the duplicate value @cstevenson1 mentioned above - https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-644323018. Once that is complete I believe the field tool related updates for this issue are complete and can move to prod.

**Comment by KCI-Ablowers** [comment]
Marking as "Push to Prod - Confirmed" but leaving open pending any feedback from DelDOT.

**Comment by kanglu666** [comment]
Duplicate value has been removed from dev.
Problem code Domain has been updated.
Value in LU_WORK_ORDER_CODES table and  WORK_ORDER_PROBLEM_CODES table have been updated.

**Comment by KCI-Ablowers** [comment]
@gerrykelly and @kanglu666 - We are good to move all of this to prod (new prod).

**Comment by gerrykelly** [comment]
@KCI-Ablowers 
Completed.

**Comment by KCI-Ablowers** [comment]
From DelDOT on 9/1/20 - 

> Hi Matthew,
> 
> All the BMP workorder has been loaded into Maximo.  Please check.
> 
> For Structure: Out of 3, 2 workorder successfully loaded. One failed as problem code did not match in Maximo,  problem code on KCI side VACBAS, in Maximo it is VACBACS.
> 
> 
> Thanks,
> Supreet
> 
 
Problem codes provided here - https://github.com/leeensminger/DelDOT-NPDES-Viewer/issues/338#issuecomment-558294516 show that the problem code should be `VACBAS` as present in the database. @cstevenson1 can you confirm that it should be `VACBACS` as Supreeet mentions above? 

![image](https://user-images.githubusercontent.com/35460336/91892615-fafb2500-ec60-11ea-91bd-87352ccec1ee.png)

**Comment by cstevenson1** [comment]
@KCI-Ablowers Please put this item on hold. The values in the spreadsheet came from DelDOT a while back so Matt and I plan to discuss internally and with DelDOT to figure out which side (KCI or DelDOT) should change the code.
